cmd/promtool/rules.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// NewRuleImporter creates a new rule importer that can be used to backfill rules.
+func NewRuleImporter(logger log.Logger, config RuleImporterConfig) *RuleImporter {
+	return &RuleImporter{
+		config:      config,
+		groupLoader: rules.FileLoader{},
+	}
+}
+
+// Init initializes the rule importer which creates a new block writer
+// and creates an Prometheus API client.
+func (importer *RuleImporter) Init() error {
+	importer.writer = blocks.NewMultiWriter(importer.logger,
+		importer.config.OutputDir,
+		importer.config.EvalInterval.Nanoseconds(),
+	)
+
+	config := api.Config{
+		Address: importer.config.URL,
+	}
+	c, err := api.NewClient(config)
+	if err != nil {
+		return err
+	}
+	importer.apiClient = v1.NewAPI(c)
+	return nil
+}
+
+// Close cleans up any open resources.
+func (importer *RuleImporter) Close() error {
+	return importer.writer.Close()
+}
+
+// LoadGroups reads groups from a list of rule files.
+func (importer *RuleImporter) LoadGroups(ctx context.Context, filenames []string) (errs []error) {
+	groups := make(map[string]*rules.Group)
+
+	for _, filename := range filenames {
+		rgs, errs := importer.groupLoader.Load(filename)
+		if errs != nil {
+			return errs
+		}
+
+		for _, ruleGroup := range rgs.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rgRules := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := importer.groupLoader.Parse(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, filename)}
+				}
+				rgRules = append(rgRules, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[rules.GroupKey(filename, ruleGroup.Name)] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     filename,
+				Interval: itv,
+				Rules:    rgRules,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return nil
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores them in new blocks.
+func (importer *RuleImporter) ImportAll(ctx context.Context) []error {
+	var errs = []error{}
+	for _, group := range importer.groups {
+		stimeWithAlignment := group.EvalTimestamp(importer.config.Start.UnixNano())
+
+		for _, r := range group.Rules() {
+			err := importer.ImportRule(ctx, r.Query().String(), stimeWithAlignment, group.Interval())
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	_, err := importer.writer.Flush()
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+// ImportRule imports the historical data for a single rule.
+func (importer *RuleImporter) ImportRule(ctx context.Context, ruleExpr string, stimeWithAlignment time.Time, internval time.Duration) error {
+	ts := stimeWithAlignment
+	appender := importer.writer.Appender()
+
+	for ts.Before(importer.config.End) {
+		currentBlockEnd := ts.Add(blockSize * time.Hour)
```

Comment:
Couldn't this produce a timestamp that's beyond the end of the block?

Reply:
yeah, let me add a check that fixes that